python/cudf/cudf/core/frame.py

```diff
@@ -1085,42 +1085,49 @@ def from_arrow(cls, data):
-                    # is specified as 'empty' and np_dtypes as 'object',
-                    # hence handling this special case to type-cast the empty
-                    # float column to str column.
-                    dtype = np_dtypes[name]
-                elif pandas_dtypes[
-                    name
-                ] == "object" and cudf.api.types.is_struct_dtype(
-                    np_dtypes[name]
-                ):
-                    # Incase of struct column, libcudf is not aware of names of
-                    # struct fields, hence renaming the struct fields is
-                    # necessary by extracting the field names from arrow
-                    # struct types.
-                    result[name] = result[name]._rename_fields(
-                        [field.name for field in data[name].type]
-                    )
+        for name in result:
+            dtype = None
+            if (
+                len(result[name]) == 0
+                and pandas_dtypes.get(name, "") == "categorical"
+            ):
+                # When pandas_dtype is a categorical column and the size
+                # of column is 0(i.e., empty) then we will have an
+                # int8 column in result._data[name] returned by libcudf,
+                # which needs to be type-casted to 'category' dtype.
+                dtype = "category"
+            elif (
+                pandas_dtypes.get(name, "") == "empty"
+                and np_dtypes.get(name, "") == "object"
+            ):
+                # When a string column has all null values, pandas_dtype is
+                # is specified as 'empty' and np_dtypes as 'object',
+                # hence handling this special case to type-cast the empty
+                # float column to str column.
+                dtype = np_dtypes[name]
+            elif name in data.column_names and isinstance(
+                data[name].type,
+                (pa.StructType, pa.ListType, pa.Decimal128Type),
+            ):
+                # Incase of struct column, libcudf is not aware of names of
+                # struct fields, hence renaming the struct fields is
+                # necessary by extracting the field names from arrow
+                # struct types.
+
+                # Incase of decimal column, libcudf is not aware of the
+                # decimal precision.
+
+                # Incase of list column, there is a possibility of nested
+                # list columns to have struct & decimal columns inside them.
+
+                # All these are taken care by calling to _with_type_metadata
+                # API on the column.
+                result[name] = result[name]._with_type_metadata(
+                    cudf.utils.dtypes.cudf_dtype_from_pa_type(data[name].type)
+                )
 
-                if dtype is not None:
-                    result[name] = result[name].astype(dtype)
+            if dtype is not None:
```

Comment:
This should be unreachable if we hit the struct/decimal/list case above, right? That is, I think it would be equivalent (and short-circuit) to write `continue` after `result[name] = result[name]._with_type_metadata(...)`. However, I am wondering if there's a better order of the conditionals here, to make the flow more natural.

Reply:
One way that comes to my mind is duplicating `result[name] = result[name].astype(dtype)` in first two if/elif conditions.

Reply:
Or, do an `as_categorical_column` in first `if` and `.astype("str")`/`.as_string_column` in `elif`? That way we can get rid of `dtype`

Reply:
> Or, do an `as_categorical_column` in first `if` and `.astype("str")`/`.as_string_column` in `elif`? That way we can get rid of `dtype` and related dead-code.

I actually went ahead and refactored with this approach, let me know if you feel old code was better I can revert it back.